ui/src/ModalFormComponents/Select.tsx

```diff
@@ -15,59 +15,143 @@
  *   limitations under the License.
  */
 
-import React, {ReactNode, useEffect, useState} from 'react';
+import React, {ReactNode, useEffect, useRef, useState} from 'react';
+import {debounce} from '../Utils';
 
 import './Select.scss';
 
+const DEFAULT_CURRENT_INDEX = 0;
+
 export interface OptionType {
     value: unknown;
+    ariaLabel: string;
     displayValue: ReactNode | string;
 }
 
 interface Props {
+    ariaLabel: string;
     selectedOption: OptionType;
     options: Array<OptionType>;
     onChange: (selectedOption: OptionType) => void;
 }
 
-const Select = ({ options, selectedOption, onChange }: Props): JSX.Element => {
+const Select = ({ ariaLabel, options, selectedOption, onChange }: Props): JSX.Element => {
     const [dropdownToggle, setDropdownToggle] = useState<boolean>(false);
     const [currentOption, setCurrentOption] = useState<OptionType>(selectedOption);
+    const dropdownToggleElement = useRef<HTMLButtonElement>(null);
+    const dropdownElement = useRef<HTMLUListElement>(null);
+    const [currentIndex, setCurrentIndex] = useState<number>(
+        DEFAULT_CURRENT_INDEX
+    );
+    const [upKey, downKey, enterKey] = [38, 40, 13];
+
+    useEffect(() => {
+        const currentOptionIndex = options.map(option => JSON.stringify(option.value))
+            .indexOf(JSON.stringify(currentOption.value));
+        setCurrentIndex(currentOptionIndex);
+    }, [options, currentOption]);
 
     useEffect(() => {
         setCurrentOption(selectedOption);
     }, [selectedOption]);
 
+    useEffect(() => {
+        const focusOnDropdown = (): void => {
+            if (dropdownToggle && !!dropdownElement.current) {
+                dropdownElement.current.focus();
+            }
+        };
+
+        focusOnDropdown();
+    }, [dropdownToggle, currentIndex]);
+
     const showDropdown = (): void => {
-        if (dropdownToggle) {
-            hideDropdown();
+        debounce(() => {
+            setDropdownToggle(true);
+        }, 100)();
+    };
+
+    const toggleDropdown = (): void => {
+        if (!dropdownToggle) {
+            showDropdown();
         } else {
-            setDropdownToggle(!dropdownToggle);
-            document.addEventListener('click', hideDropdown, false);
+            hideDropdown();
         }
     };
 
     const hideDropdown = (): void => {
-        setDropdownToggle(false);
-        document.removeEventListener('click', hideDropdown);
+        debounce(() => {
+            setDropdownToggle(false);
+        }, 100)();
+    };
+
+    const setSelectedItem = (index: number): void => {
+        if (currentIndex !== index) {
+            setCurrentIndex(index);
+        }
+    };
+
+    const updateSelectedOption = (option: OptionType, currentIndex: number): void => {
+        setCurrentOption(option);
+        setCurrentIndex(currentIndex);
```

Comment:
This will make two re-renders happen sequentially. Consider consolidating into a single state object?

Reply:
That's a good point, but they aren't always in sync because the current index is more for tracking which color to highlight when a user is using the up and down arrows, whereas the currentOption is the actual option the user has already selected.